cmd/statreceiver/influx.go

```diff
@@ -0,0 +1,152 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/zeebo/errs"
+)
+
+// InfluxDest is a MetricDest that sends data with the Influx TCP wire
+// protocol
+type InfluxDest struct {
+	url   string
+	token string
+
+	mu      sync.Mutex
+	buf     bytes.Buffer
+	stopped bool
+}
+
+// NewInfluxDest creates a InfluxDest with stats URL url. Because
+// this function is called in a Lua pipeline domain-specific language, the DSL
+// wants a Influx destination to be flushing every few seconds, so this
+// constructor will start that process. Use Close to stop it.
+func NewInfluxDest(writeURL string) *InfluxDest {
+	parsed, err := url.Parse(writeURL)
+	if err != nil {
+		panic(err)
+	}
+	token := parsed.Query().Get("authorization")
+	parsed.Query().Del("authorization")
+
+	rv := &InfluxDest{
+		url:   parsed.String(),
+		token: token,
+	}
+	go rv.flush()
+	return rv
+}
+
+// Metric implements MetricDest
+func (d *InfluxDest) Metric(application, instance string, key []byte, val float64, ts time.Time) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	// TODO(jeff): actual parsing of the key is very tricky in the presence of influx's busted
+	// escapes. if we could do that, we could more easily put the application tag in sorted order
+	// but since it begins with a, we'll do the easy thing and insert it first.
+	added := false
+keyRange:
+	for i, val := range key {
+		switch {
+		case val != ' ' && val != ',':
+			continue
+		case i == 0:
+			break keyRange
+		case key[i-1] == '\\':
+			continue
+		}
+
+		var newKey []byte
+		newKey = append(newKey, key[:i]...)
+		newKey = append(newKey, ",application="...)
+		newKey = appendTag(newKey, application)
+		newKey = append(newKey, key[i:]...)
+		key = newKey
+
+		added = true
+		break
+	}
+	if !added {
+		// TODO(jeff): log that it was dropped? how?
```

Comment:
turns out later on we just use `log.Printf` which i got from other files, so we can do that here.